### PR TITLE
Do not break pipeline for window queries with GROUPS

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#5586](https://github.com/apache/arrow-datafusion/issues/5586).

# Rationale for this change

Currently, when window queries involve `GROUPS` window frame type. We cannot run it using `BoundedWindowAggExec`. Instead we use `WindowAggExec`. This breaks pipeline, even if there is no theoretical reason to do so. With this change we solve this problem. See the [#5586](https://github.com/apache/arrow-datafusion/issues/5586) for better explanation.

# What changes are included in this PR?

With this PR physical plan of the query
```sql
SELECT
    SUM(c6) OVER(ORDER BY c2 GROUPS BETWEEN 1 PRECEDING AND 1 FOLLOWING) as sum1
    FROM aggregate_test_100
```
would turn from
```sql
"ProjectionExec: expr=[SUM(aggregate_test_100.c6) ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] GROUPS BETWEEN 1 PRECEDING AND 1 FOLLOWING@13 as sum1]",
"  WindowAggExec: wdw=[SUM(aggregate_test_100.c6): Ok(Field { name: \"SUM(aggregate_test_100.c6)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Groups, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
"    SortExec: expr=[c2@1 ASC NULLS LAST]",
"      CsvExec: files={1 group: [[file_path]]}, has_header=true, limit=None, projection=[c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13]",
```
to
```sql
"ProjectionExec: expr=[SUM(aggregate_test_100.c6) ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] GROUPS BETWEEN 1 PRECEDING AND 1 FOLLOWING@13 as sum1]",
"  BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c6): Ok(Field { name: \"SUM(aggregate_test_100.c6)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Groups, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
"    SortExec: expr=[c2@1 ASC NULLS LAST]",
"      CsvExec: files={1 group: [[file_path]]}, has_header=true, limit=None, projection=[c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13]",
```
 This enables us to not break the pipeline when window queries involve `GROUPS` clause (given that there is no other pipeline-breaking executor in the plan). 

# Are these changes tested?
Yes. Fuzzy tests are extended to cover this use case. Also end to end tests are added.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 1]
```diff
@@ -548,9 +548,9 @@ impl SortedPartitionByBoundedWindowStream {
             for (partition_row, WindowState { state: value, .. }) in window_agg_state {
                 let n_prune =
                     min(value.window_frame_range.start, value.last_calculated_index);
-                if let Some(state) = n_prune_each_partition.get_mut(partition_row) {
-                    if n_prune < *state {
-                        *state = n_prune;
+                if let Some(current) = n_prune_each_partition.get_mut(partition_row) {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:551]: The changes in this file appear to be a drive by cleanup (as in not directly related to this PR) -- that is fine I am just confirming my understanding

File: datafusion/core/tests/window_fuzz.rs [hunk 2]
```diff
@@ -227,9 +227,7 @@ fn get_random_window_frame(rng: &mut StdRng) -> WindowFrame {
     } else if rand_num < 2 {
         WindowFrameUnits::Rows
     } else {
-        // For now we do not support GROUPS in BoundedWindowAggExec implementation
-        // TODO: once GROUPS handling is available, use WindowFrameUnits::GROUPS in randomized tests also.
-        WindowFrameUnits::Range
+        WindowFrameUnits::Groups
```
Comment (alamb) [inline datafusion/core/tests/window_fuzz.rs:230]: 👍

File: datafusion/physical-expr/src/window/partition_evaluator.rs [hunk 3]
```diff
@@ -38,9 +38,15 @@ pub trait PartitionEvaluator: Debug + Send {
         Ok(BuiltinWindowState::Default)
     }
 
+    /// Updates the internal state for Built-in window function
```
Comment (alamb) [inline datafusion/physical-expr/src/window/partition_evaluator.rs:41]: thank you for adding these comments

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @mustafasrepo  and @ozankabak

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs (hunk 1 above)
Comment (ozankabak) [reply datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:551]: Right, this is a drive-by cleanup. There were too many `state`s going around making it hard to read 🙂

**Pull request merged**